### PR TITLE
Provide consistent interface for Synthetics Tests API

### PR DESCRIPTION
**Pull request by danielfurman**
GetTestResults() and GetTestTrace() methods will be implemented in next PR.
    
Based-on: https://github.com/kentik/community_sdk_golang/pull/104
Issue: KNTK-399

### CODE REVIEW

File: examples/synthetics_example_test.go
Comment (mmac-m3a) [inline examples/synthetics_example_test.go]: If this is always executed against a "fake" server, then hardcoding agent IDs is fine. Otherwise, agent IDs need to be selected by matching available agent by some criteria.

File: examples/synthetics_example_test.go
Comment (mmac-m3a) [inline examples/synthetics_example_test.go:413]: Is the `port` set somewhere to a usable value?

File: examples/synthetics_example_test.go
Comment (mmac-m3a) [inline examples/synthetics_example_test.go:416]: Same here.

File: examples/synthetics_example_test.go
Comment (mmac-m3a) [inline examples/synthetics_example_test.go]: Not very realistic port for DNS (which defaults to 53).

File: examples/synthetics_example_test.go
Comment (mmac-m3a) [inline examples/synthetics_example_test.go]: Same here.

File: kentikapi/internal/cloud/cloud.go
Comment (mmac-m3a) [inline kentikapi/internal/cloud/cloud.go]: More precisely "cloudexports" API.

File: kentikapi/internal/synthetics/test.go
Comment (mmac-m3a) [inline kentikapi/internal/synthetics/test.go:120]: There is also `bgp_monitor` test type which should be silently ignored (it will be handled in separate BGP monitoring API).

Comment (mmac-m3a) [reply kentikapi/internal/synthetics/test.go:120]: The API may already return `transaction` type test, but it cannot yet provide actual configuration or results.

File: kentikapi/internal/synthetics/test.go
Comment (mmac-m3a) [inline kentikapi/internal/synthetics/test.go:147]: Can go support general (introspection based) [construction of SDK objects based on protobuf data](https://github.com/kentik/community_sdk_python/blob/1bac5553fc04ba3d536602a372cd380c3ba32f87/kentik_api_library/kentik_api/synthetics/synth_tests/base.py#L56) like Python does?

File: kentikapi/internal/synthetics/test.go
Comment (mmac-m3a) [inline kentikapi/internal/synthetics/test.go]: We need to support the ability to provide `edate` in `TestUpdate` requests.  If provided, the API returns error if the object to be modified has been modified in the DB after the `edate` timestamp. This allows to guard against concurrent modifications.

File: kentikapi/internal/synthetics/test.go
Comment (mmac-m3a) [inline kentikapi/internal/synthetics/test.go:437]: Tests now also support `labels`. Either add the support now, or create a ticket for later.

File: kentikapi/synthetics/test.go
Comment (mmac-m3a) [inline kentikapi/synthetics/test.go]: Typo: "lst" -> "last"

**Review by mmac-m3a — APPROVED** [review]
I'd like to get your feedback to my comments before merging the code. The code is good to go in as long as we've tickets for:
- support for labels for tests and agents
- support for passing `edate` in `TestUpdate` requests

File: kentikapi/internal/cloud/export.go
Comment (mateuszmidor) [inline kentikapi/internal/cloud/export.go]: As the list of supported providers may grow and cause error, maybe "unsupported cloud provider" woud better describe the case

File: kentikapi/internal/convert/convert.go
Comment (mateuszmidor) [inline kentikapi/internal/convert/convert.go]: nanosPerMilli, probably

**Review by mateuszmidor — COMMENTED** [review]
Looks good, minor comments left

File: examples/synthetics_example_test.go
Comment (danielfurman) [reply examples/synthetics_example_test.go]: Done: 2e37189
Examples are always executed against production server. Added note about that in readme and modified code to randomly select agents based on criteria.

File: examples/synthetics_example_test.go
Comment (danielfurman) [reply examples/synthetics_example_test.go:413]: No, it was not. I added it to required fields and filled with value for TCP pings. For ICMP pings this field can still be omitted, because 0 is zero value for uint32.
89dcddc

File: examples/synthetics_example_test.go
Comment (danielfurman) [reply examples/synthetics_example_test.go:416]: I wrongly assumed that server applies default here.
I added default value 33434 for UDP and TCP trace, which hopefully makes sense.
f1827f5

File: examples/synthetics_example_test.go
Comment (danielfurman) [reply examples/synthetics_example_test.go]: Yeah, I didn't bother with realism too much. Changed to 53.
eba3175

File: examples/synthetics_example_test.go
Comment (danielfurman) [reply examples/synthetics_example_test.go]: Done: https://github.com/kentik/community_sdk_golang/commit/eba3175dbd258b7b6da6678f4ae4d95f5c430de4

File: kentikapi/internal/cloud/cloud.go
Comment (danielfurman) [reply kentikapi/internal/cloud/cloud.go]: Fixed: e741aa1

File: kentikapi/internal/synthetics/test.go
Comment (danielfurman) [reply kentikapi/internal/synthetics/test.go:120]: Added code for bgp_monitor tests: 4e80fa9

Comment (danielfurman) [reply kentikapi/internal/synthetics/test.go:120]: Added code for transaction tests: 3d7e8d2

**Comment by mmac-m3a** [comment]
`godot` days: `Error: Comment should end in a period (godot)`... Like anybody was [waiting for him](https://en.wikipedia.org/wiki/Waiting_for_Godot) to days that :-).

File: examples/synthetics_example_test.go
Comment (mmac-m3a) [inline examples/synthetics_example_test.go]: That should be printed in large blinking red, if the value is non-zero :). The number is actually of any use for Kentik engineers. A non-zero value indicates some bug in the Kentik system. Let's print it only if it is non-zero and add text suggesting to contact Kentik support. Something like: `Kentik API returned <N> invalid agents. Please, contact Kentik support.`.

File: examples/synthetics_example_test.go
Comment (mmac-m3a) [inline examples/synthetics_example_test.go:80]: Why do you insist on using a private agent here? There are definitely synthetics customers that have no private agents and still may want to learn about the SDK by running/tweaking examples.

Using a public agent IMHO makes more sense. There is (almost) guaranteed to be at least one available and it allows Kentik Trial customers to easily test the Go SDK.

File: examples/synthetics_example_test.go
Comment (mmac-m3a) [inline examples/synthetics_example_test.go]: Like for agents, this number is of any use only to Kentik engineers.

File: examples/synthetics_example_test.go
Comment (mmac-m3a) [inline examples/synthetics_example_test.go]: Considering that the goal of the example is to demonstrate the Go SDK (as opposed to test the Kentik API), let's use only IPv4 agents and tests. The reason for this is that the API is little bit behind the development of the Kentik synthetics system, and it currently cannot reliably represent dual stacked agents (it has no place to show the "other" IP address. Which one (v4 or v6) of addresses is shown in the `ip` (and also `local_ip`) property is undefined. The system internally (since about 2 months ago) does the right thing, but it is not visible in the API. The visible behavior may confuse more detailed oriented customers :).

We can add examples of choosing dual stack agents and picking protocol family in the test once the API catches up and we add support for new fields in the SDK.

File: examples/synthetics_example_test.go
Comment (mmac-m3a) [inline examples/synthetics_example_test.go:577]: Expecially for private agents, you need to also test that the `agent.status` is OK. If you want to be increase probability that the agent is actually operational, you can add test on the `last_authed` timestamp being less than 25 hours in the past (agents re-authorize every 24 hours).
Note that picking a non-operational agent causes test results to be unavailable, otherwise test creation (unfortunately) succeeds.

**Review by mmac-m3a — COMMENTED** [review]
Getting ever better.

File: examples/synthetics_example_test.go
Comment (mmac-m3a) [reply examples/synthetics_example_test.go]: Note that soon (PR is pending review) node.js (`ksynth-agent`) agents will be allowed to run any synthetic test. Rust (`ksynth`) agents will never be able to run `page-load` and `transaction` tests.

File: kentikapi/internal/synthetics/test.go
Comment (danielfurman) [reply kentikapi/internal/synthetics/test.go:147]: Yes, I think it is possible, using the `reflect` package. There are several caveats, however:
1. I renamed some fields to be more user-friendly. That could be reverted or we could use struct tags on public models to specify PB field name.
2. Some PB fields are translated to user-friendly types, e.g. `string` to `ip.Addr`, `string` with number to `int`, `string` with duration to `time.Duration`. That could be reverted or we could use struct tags on public models to specify custom translation function.
3. I merged `activation.time_unit` and `activation.time_window` PB fields to single `Activation.TimeWindow` field of type time.Duration. That could be reverted or we could find some other mechanism to keep that state.
4. Possibly other aspects
5. There is a library for copying fields from one struct to other, but it might not be flexible enough: https://github.com/jinzhu/copier/  Its custom converters look messy: https://github.com/jinzhu/copier/pull/115

It would be quite hard to write and read. It might be more maintainable in the long run, but I am skeptical. I definitely don't want to do that in this PR. Do you want me to create a task for that?

File: kentikapi/internal/synthetics/test.go
Comment (danielfurman) [reply kentikapi/internal/synthetics/test.go]: Done: faa5fc9

File: kentikapi/synthetics/test.go
Comment (danielfurman) [reply kentikapi/synthetics/test.go]: Done https://github.com/kentik/community_sdk_golang/commit/faa5fc9cb5bf42f91e09e96df25f8e05d5727bf2

File: kentikapi/internal/synthetics/test.go
Comment (danielfurman) [reply kentikapi/internal/synthetics/test.go:437]: Task created: https://clkentik.atlassian.net/browse/KNTK-409
It is not possible to do it now without manual modifications of generated proto structs, which we want to avoid.

File: kentikapi/internal/cloud/export.go
Comment (danielfurman) [reply kentikapi/internal/cloud/export.go]: Done: d5528a9

File: kentikapi/internal/convert/convert.go
Comment (danielfurman) [reply kentikapi/internal/convert/convert.go]: Nice catch; fixed

File: examples/synthetics_example_test.go
Comment (danielfurman) [reply examples/synthetics_example_test.go]: Done: 061b51c
Skipped the blinking red part ;)

File: examples/synthetics_example_test.go
Comment (danielfurman) [reply examples/synthetics_example_test.go:80]: For public agents, update and deactivate methods fail with `client.Synthetics.Deactivate: code: NotFound, message: rpc error: code = NotFound desc = could not find record (name:CustomError type:object) (errxid caxg9bnwhgf46mtema0g)`

From my point of view, the more methods are shown in the example, the better - both for users and maintainers. Especially as I am treating those examples as acceptance/integration tests.

I added a comment with instructions how to run the example without private agent: 2d46810 Is it a sufficient compromise?

File: examples/synthetics_example_test.go
Comment (danielfurman) [reply examples/synthetics_example_test.go]: Done: e9566f4

File: examples/synthetics_example_test.go
Comment (danielfurman) [reply examples/synthetics_example_test.go]: And fixed again: e9566f4

File: examples/synthetics_example_test.go
Comment (danielfurman) [reply examples/synthetics_example_test.go]: Modified example to use IPv4 agents: 695420c
Let us know when changes to agents arrive: we'll update docstrings and simplify `TestDemonstrateSyntheticsTestsAPI_CreateMinimalTests`

File: examples/synthetics_example_test.go
Comment (danielfurman) [reply examples/synthetics_example_test.go:577]: Ok, but this example (`demonstrateSyntheticsAgentsAPI`) only shows available methods of Agents API. It's purpose is not finding working agents, display test results, etc.

The same for examples for synth tests (`TestDemonstrateSyntheticsTestsAPI`): they only do some API manipulation and immediately clean up after.

**Comment by danielfurman** [comment]
@mateuszmidor @mmac-m3a All comments addressed

File: kentikapi/internal/synthetics/test.go
Comment (mmac-m3a) [reply kentikapi/internal/synthetics/test.go:147]: Thanks for the explanation. We do not need to do it if the benefits are comparatively small.

File: examples/synthetics_example_test.go
Comment (mmac-m3a) [reply examples/synthetics_example_test.go:80]: Yes. I've missed that you are demonstrating agent activation and deactivation. Private agent is definitely needed for that.

File: examples/synthetics_example_test.go
Comment (mmac-m3a) [reply examples/synthetics_example_test.go:577]: OK.

**Review by mmac-m3a — APPROVED** [review]

**Pull request merged**
